### PR TITLE
*: use independent lock

### PR DESCRIPTION
**Pull request by rleungx**
Signed-off-by: Ryan Leung <rleungx@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #5586.

### What is changed and how does it work?

This PR split the basic cluster lock into two and move out some logic from the lock section in the region heartbeat.

```commit-message
```

### Check List

Tests

- Manual test

Test via 1000K Region/40 Store/1PD

**[8e18727](https://github.com/tikv/pd/pull/5587/commits/8e1872731c6ab516a971580bc1a745b37dd03588)**
```
Update speed of each category:
  Requests/sec:     41370.8813
  Save-Tree/sec:     2482.2529
  Save-KV/sec:       1654.8353
  Save-Space/sec:    6205.6322
  Save-Flow/sec:    14479.8085
  Skip/sec:         20627.0333
```

**[7d128cb](https://github.com/tikv/pd/pull/5587/commits/7d128cbfdb37fbb4dccaf1e57c74eda8f0553a63)**
```
Update speed of each category:
  Requests/sec:     34475.6364
  Save-Tree/sec:     2068.5382
  Save-KV/sec:       1379.0255
  Save-Space/sec:    5171.3455
  Save-Flow/sec:    12066.4727
  Skip/sec:         17189.1455
```
**master:**
```
Update speed of each category:
  Requests/sec:     25145.4940
  Save-Tree/sec:     1508.7296
  Save-KV/sec:       1005.8198
  Save-Space/sec:    3771.8241
  Save-Flow/sec:     8800.9229
  Skip/sec:         12537.2466
```

### Release note

```release-note
None
```

### CODE REVIEW

File: server/core/basic_cluster.go [hunk 1]
```diff
@@ -313,54 +399,29 @@ func (bc *BasicCluster) getWriteRate(
 
 // GetStoresLeaderWriteRate get total write rate of each store's leaders.
 func (bc *BasicCluster) GetStoresLeaderWriteRate() (storeIDs []uint64, bytesRates, keysRates []float64) {
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
 	return bc.getWriteRate(bc.Regions.GetStoreLeaderWriteRate)
```
Comment (bufferflies) [inline server/core/basic_cluster.go:404]: the region lock can be release earlier?

File: server/core/basic_cluster.go [hunk 2]
```diff
@@ -115,188 +102,288 @@ func (bc *BasicCluster) GetFollowerStores(region *RegionInfo) []*StoreInfo {
...
+	defer bc.Regions.mu.RUnlock()
+	return bc.Regions.GetStoreRegions(storeID)
+}
+
+// GetLeaderStoreByRegionID returns the leader store of the given region.
+func (bc *BasicCluster) GetLeaderStoreByRegionID(regionID uint64) *StoreInfo {
+	bc.Regions.mu.RLock()
+	region := bc.Regions.GetRegion(regionID)
+	if region == nil || region.GetLeader() == nil {
+		return nil
```
Comment (bufferflies) [inline server/core/basic_cluster.go:230]: need to release region RLock？

Comment (rleungx) [reply server/core/basic_cluster.go:230]: nice catch

File: server/core/basic_cluster.go [hunk 3]
```diff
@@ -115,188 +102,289 @@ func (bc *BasicCluster) GetFollowerStores(region *RegionInfo) []*StoreInfo {
...
+// GetStoreRegions gets all RegionInfo with a given storeID.
+func (bc *BasicCluster) GetStoreRegions(storeID uint64) []*RegionInfo {
+	bc.Regions.mu.RLock()
+	defer bc.Regions.mu.RUnlock()
+	return bc.Regions.GetStoreRegions(storeID)
+}
+
+// GetLeaderStoreByRegionID returns the leader store of the given region.
+func (bc *BasicCluster) GetLeaderStoreByRegionID(regionID uint64) *StoreInfo {
+	bc.Regions.mu.RLock()
```
Comment (bufferflies) [inline server/core/basic_cluster.go:227]: Is it better to use the function `GetRegion(regionID uint64)` to decrease the same code?

**Comment by nolouch** [comment]
PTAL @CabinfeverB

File: server/core/basic_cluster.go [hunk 4]
```diff
@@ -313,54 +401,31 @@ func (bc *BasicCluster) getWriteRate(
 
 // GetStoresLeaderWriteRate get total write rate of each store's leaders.
 func (bc *BasicCluster) GetStoresLeaderWriteRate() (storeIDs []uint64, bytesRates, keysRates []float64) {
-	return bc.getWriteRate(bc.Regions.GetStoreLeaderWriteRate)
+	bc.Regions.mu.RLock()
```
Comment (bufferflies) [inline server/core/basic_cluster.go:404]: the result of  storeLeaderWriteRate can not be accurate unless  lock or not lock.

File: server/core/region.go [hunk 5]
```diff
@@ -1293,6 +1293,16 @@ func isInvolved(region *RegionInfo, startKey, endKey []byte) bool {
 	return bytes.Compare(region.GetStartKey(), startKey) >= 0 && (len(endKey) == 0 || (len(region.GetEndKey()) > 0 && bytes.Compare(region.GetEndKey(), endKey) <= 0))
 }
 
+func isRegionRecreated(region *RegionInfo) bool {
```
Comment (bufferflies) [inline server/core/region.go:1296]:
```suggestion
func (region *RegionInfo)isRegionRecreated bool 
```

File: server/core/basic_cluster.go [hunk 6]
```diff
@@ -412,58 +463,68 @@ func (bc *BasicCluster) CheckAndPutRegion(region *RegionInfo) []*RegionInfo {
 	return bc.PutRegion(region)
 }
 
+// AtomicCheckAndPutRegion checks if the region is valid to put, if valid then put.
+func (bc *BasicCluster) AtomicCheckAndPutRegion(region *RegionInfo) ([]*RegionInfo, error) {
```
Comment (bufferflies) [inline server/core/basic_cluster.go:467]: what's the difference with PreCheckPutRegion？ only setRegion？

Comment (rleungx) [reply server/core/basic_cluster.go:463]: Yes.

Comment (bufferflies) [reply server/core/basic_cluster.go:463]: how about decrease some same code？

File: server/core/basic_cluster.go [hunk 7]
```diff
@@ -115,195 +102,298 @@ func (bc *BasicCluster) GetFollowerStores(region *RegionInfo) []*StoreInfo {
...
+}
+
+// DeleteStore deletes a store.
+func (bc *BasicCluster) DeleteStore(store *StoreInfo) {
+	bc.Stores.mu.Lock()
+	defer bc.Stores.mu.Unlock()
+	bc.Stores.DeleteStore(store)
+}
+
+/* Regions read operations */
```
Comment (bufferflies) [inline server/core/basic_cluster.go:195]: Delete this line?

File: server/core/basic_cluster.go [hunk 8]
```diff
@@ -115,195 +102,298 @@ func (bc *BasicCluster) GetFollowerStores(region *RegionInfo) []*StoreInfo {
...
 	f func(storeID uint64) (bytesRate, keysRate float64),
 ) (storeIDs []uint64, bytesRates, keysRates []float64) {
-	bc.RLock()
-	defer bc.RUnlock()
+	bc.Stores.mu.RLock()
+	defer bc.Stores.mu.RUnlock()
 	count := len(bc.Stores.stores)
 	storeIDs = make([]uint64, 0, count)
 	bytesRates = make([]float64, 0, count)
 	keysRates = make([]float64, 0, count)
```
Comment (bufferflies) [inline server/core/basic_cluster.go:391]: it only needs to get all  store ids by using this store read lock. so it can release it earlier

**Review by bufferflies — COMMENTED** [review]
rest LGTM

**Review by bufferflies — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**